fix a bug when splitting at a meta2 entry

tbg opened the pull request:
fixes a bug which caused a panic when clearing overlapping descriptors for a
range `[KeyMin, Meta2Key)`. The issue was that when attempting to clear out
descriptors which were subsumed by the above range, an iteration over the
range given by its meta keys was performed, with the first endpoint excluded.
This exclusion was performed incorrectly by first incrementing the start key
(`KeyMin`) and then forming the meta key; for `KeyMin` this leads to
`Meta2Prefix\x00`. For the above `EndKey`, the meta key is a Meta1key which sorts
before `Meta2Prefix\x00`, causing a panic. The fix was simply to increment the
meta key for `StartKey`, not `StartKey` itself.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=1778) <br/>You should sign our Contributor License Agreement in order to get your pull request merged.<br/>

MBerhault commented:
LGTM

tamird commented on kv/range_cache_test.go:
```diff
@@ -262,3 +262,45 @@ func TestRangeCacheClearOverlapping(t *testing.T) {
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				t.Fatalf("invocation of clearOverlappingCachedRangeDescriptors panic'ed")
```
panicked?

tamird replied:
i'd log the panic too

tschottdorf replied:
done.

tamird commented:
LGTM

tbg committed:
fix a bug when splitting at a meta2 entry

fixes a bug which caused a panic when clearing overlapping descriptors for a
range [KeyMin, Meta2Key). The issue was that when attempting to clear out
descriptors which were sub- sumed by the above range, an iteration over the
range given by its meta keys was performed, with the first endpoint excluded.
This exclusion was performed incorrectly by first incrementing the start key
(KeyMin) and then forming the meta key; for KeyMin this leads to
Meta2Prefix\x00. For the above EndKey, the meta key is a Meta1key which sorts
before Meta2Prefix\x00, causing a panic. The fix was simply to increment the
meta key for StartKey, not StartKey itself.

tbg committed:
fix variable names

spencerkimball commented on kv/range_cache.go:
```diff
@@ -255,12 +255,12 @@ func (rdc *rangeDescriptorCache) clearOverlappingCachedRangeDescriptors(key, met
 	// Also clear any descriptors which are subsumed by the one we're
 	// going to cache. This could happen on a merge (and also happens
-	// when there's a lot of concurrency). Iterate from StartKey.Next().
+	// when there's a lot of concurrency). Iterate from after StartKey.
```
How about:

Iterate from the range meta key after RangeMetaKey(desc.StartKey) to the range meta key for desc.EndKey.

tschottdorf replied:
done.

spencerkimball commented:
LGTM
